### PR TITLE
Replace icon system

### PR DESCRIPTION
**Pull request by dslemay**
https://github.com/phase2/particle/issues/316

### CODE REVIEW

**Comment by yuginaili** [comment]
@dslemay PR looks great! Is the old font icon code commented out just for reference or can we yank it out?

**Comment by dslemay** [comment]
@yuginaili Thank you. Eventually it will be able to be ripped out. There is still more work to be done on this that I will be focusing on today including: checking IE, and creating a demo atom with accompanying documentation on how to use it.

File: .gitignore [hunk 1]
```diff
@@ -15,6 +17,7 @@ node_modules
 vendor
 
 # Generated files
+source/_data/scssVariables.json
```
Comment (illepic) [inline .gitignore]: Move this up to the other `source/` files, not down here with os generated files.

File: apps/pl/scss/_scss2json.scss [hunk 2]
```diff
@@ -29,13 +29,15 @@ $spacing-demo: $spacing;
 //@import '../_patterns/01-atoms/button/_button.scss';
 
 // Import the map of available custom font icons.
-@import '../../../source/_patterns/01-atoms/icon/scss/_icons-generated';
+@import '../../../source/_patterns/01-atoms/svgicon/scss/_icons-generated';
```
Comment (illepic) [inline apps/pl/scss/_scss2json.scss]: Move `@import`s to top of file, generally.

File: apps/pl/scss/_scss2json.scss [hunk 3]
```diff
@@ -29,13 +29,15 @@ $spacing-demo: $spacing;
...
 
-$icons-demo: (
-  prefix: 'icon--',
-  fontname: 'iconfont',
-  icons: map-keys(map_get($__iconfont__data, 'iconfont')),
-);
+$svgicons: map-keys($sprites);
```
Comment (illepic) [inline apps/pl/scss/_scss2json.scss]: Give us a comment here on what the format of this looks like.

File: apps/pl/scss/_scss2json.scss [hunk 4]
```diff
@@ -29,13 +29,15 @@ $spacing-demo: $spacing;
...
+
+// $icons-demo: (
```
Comment (illepic) [inline apps/pl/scss/_scss2json.scss]: Do we want to nuke this?

File: source/_patterns/01-atoms/svgicon/_svgicon.twig [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{#
+/**
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/_svgicon.twig]: Let's not use the js comment header in Twig. It looks nice, but let's just use what's in other twig files for now.

File: source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss [hunk 6]
```diff
@@ -0,0 +1,15 @@
+// All component Sass needs non-printing base config
+@import '../../../00-protons/variables';
+@import '../scss/_icons-generated';
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss]: Double-double verifying that this has **no** printing css in it.

File: source/_patterns/01-atoms/svgicon/demo/index.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+// Import component assets
+import 'atoms/svgicon';
+
+// Demo-only styles
+import './_svgicon-demo.scss';
+
+// Import demo assets
+import twig from './svgicons.twig';
+import yaml from './svgicons.yml';
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/index.js]: Should be in `demo/`.

File: source/_patterns/01-atoms/svgicon/demo/index.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import markdown from './svgicons.md';
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/index.js]: Should be in `demo/`.

File: source/_patterns/01-atoms/svgicon/demo/svgicons.twig [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    <pre class="language-css">
+      <code>
+        .svgicon-demo-0 {
+          @include sprite('chevron-right');
+          height: 40px;
+          width: 40px;
+        }
+      </code>
+    </pre>
+      {% for icon in scssIcons %}
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/svgicons.twig]: This needs a parent wrapper.

File: source/_patterns/01-atoms/svgicon/_svgicon.scss [hunk 10]
```diff
@@ -0,0 +1,14 @@
+// All component Sass needs non-printing base config
+@import '../../00-protons/variables';
+@import '../../../../apps/pl/scss/_scss2json';
+@import './scss/_icons-generated';
+
+// Loop over every SVG and create a CSS class for it
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/_svgicon.scss]: Let's also generate all the `.color-fill--{colorname}` classes here as well, allowing us to colorize the SVG files by class name from the design system.

File: source/_patterns/01-atoms/svgicon/demo/svgicons.twig [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+      </code>
+    </pre>
+    <div class="svgicon-demo-1">
+    </div>
+    <h2>Using Twig to render SVG elements</h2>
+    <p>Using the Twig template renders an SVG element directly to the DOM. The template accepts several arguments: <code>svgname</code> accepts the filename of the SVG to render, <code>svgfill</code> accepts a value to set the change the default fill color of the SVG element from black.</p>
+    {% include '@atoms/_svgicon.twig' with {svgname: 'chevron-up'} %}
+    {% include '@atoms/_svgicon.twig' with {svgname: 'chevron-up', svgfill: '#F00'} %}
+
+  {% endblock column_1 %}
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/svgicons.twig]: Need to demonstrate the "color svg icon by color name from design system" class concept here.

File: apps/pl/scss/_scss2json.scss (hunk 4 above)
Comment (dslemay) [reply apps/pl/scss/_scss2json.scss]: We need the svgicons variable to pass along to the JSON for the demo as well as the mapping function to create the CSS classes. We can delete the old icons-demo along with ripping out the old component.

File: apps/pl/scss/_scss2json.scss (hunk 2 above)
Comment (dslemay) [reply apps/pl/scss/_scss2json.scss]: Leaving inline per pattern throughout this particular file.

File: source/_patterns/01-atoms/svgicon/_svgicon.twig (hunk 5 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/_svgicon.twig]: New default behavior with generator.

File: source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss (hunk 6 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss]: Protons is non-printing and icons-generated contains sass variables and methods to use the mixin.

File: source/_patterns/01-atoms/svgicon/demo/index.js (hunk 7 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/demo/index.js]: In demo.

File: source/_patterns/01-atoms/svgicon/demo/index.js (hunk 8 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/demo/index.js]: In demo.

File: apps/pl/scss/_scss2json.scss (hunk 3 above)
Comment (dslemay) [reply apps/pl/scss/_scss2json.scss]: Added in 4c7aedc

File: .gitignore (hunk 1 above)
Comment (dslemay) [reply .gitignore]: Moved in 4c7aedc

File: source/_patterns/01-atoms/svgicon/demo/svgicons.twig (hunk 9 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/demo/svgicons.twig]: Added in 4c7aedc

File: source/_patterns/01-atoms/svgicon/_svgicon.scss (hunk 10 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/_svgicon.scss]: Added in c4c164c.

The twig svgfill now accepts one of the keys defined in $theme-colors as an argument which applies the related BEM class for colorizing the SVG.

File: source/_patterns/01-atoms/svgicon/demo/svgicons.twig (hunk 11 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/demo/svgicons.twig]: Functionality to use the color name from the design system was added in c4c164c. Further documentation to follow.

File: source/_patterns/01-atoms/icon/scss/_icon.scss [hunk 12]
```diff
@@ -1,53 +0,0 @@
...
-  font-style: normal;
-}
-
-@each $name, $value in (map_get($__iconfont__data, 'iconfont')) {
-  .icon--#{$name} {
-    @include icon($name);
-  }
-}
-
-// Font-Awesome Sass Examples
```
Comment (tjheffner) [inline source/_patterns/01-atoms/icon/scss/_icon.scss]: if we are still including font-awesome in the build (i think we are? didn't see it removed in this pr) can we still include the font awesome examples in the updated demo? i know many projects still use FA regularly for ease of use as an icon solution.

also consider bumping up from 4.7 to 5.0, which gives all of their new accessibility goodies for fonts

File: apps/pl/scss/_scss2json.scss (hunk 2 above)
Comment (tjheffner) [reply apps/pl/scss/_scss2json.scss]: i think this one can move to the top, the icons won't conflict with any of bootstraps stuff. the other lower import for "customized variables" is specifically in the case that we _do_ override any bootstrap values (still could probably be moved to the top, underneath the main protons import)

File: package.json [hunk 13]
```diff
@@ -90,6 +91,7 @@
     "stylelint-config-prettier": "^3.2.0",
     "stylelint-scss": "^3.0.0",
     "stylelint-webpack-plugin": "^0.10.5",
+    "svg-spritemap-webpack-plugin": "^2.6.0",
```
Comment (tjheffner) [inline package.json]: also uninstall "iconfont-plugin-webpack" if it's no longer needed for our iconfonts

File: source/_patterns/01-atoms/icon/scss/_icon.scss (hunk 12 above)
Comment (dslemay) [reply source/_patterns/01-atoms/icon/scss/_icon.scss]: This brings up a larger conversation on if we continue to use the Font Awesome library as an icon font or look to move it to their SVG spritesheet that they also provide. Thoughts? Copying over icon font for Font Awesome would be faster. Not sure I would have time to write the implementation with their SVG spritesheet before starting client work next week.

File: package.json (hunk 13 above)
Comment (dslemay) [reply package.json]: Good catch. Removed in c1cdb7f

File: apps/pl/scss/_scss2json.scss (hunk 2 above)
Comment (dslemay) [reply apps/pl/scss/_scss2json.scss]: Moved to just below the bootstrap import and also cleaned up some commented out code from the old iconfont system as well in e22f35b.

File: source/_patterns/01-atoms/icon/scss/_icon.scss (hunk 12 above)
Comment (tjheffner) [reply source/_patterns/01-atoms/icon/scss/_icon.scss]: personally i am in favor of moving to 5.0 and using their default svg + js, but i'm not opposed to spritesheets for consistency's sake. that upgrade doesn't have to happen in this PR, but at the very least i would still like to see the 4.7 examples included as-is

Comment (dslemay) [reply source/_patterns/01-atoms/icon/scss/_icon.scss]: I added the documentation of the current FA mixin system back into the demo docs for the new atom. 423dc1d

I agree that there is room for further development in upgrading FA in a future branch. The SVG + JS approach looks nice.

File: apps/pl/scss/_scss2json.scss [hunk 14]
```diff
@@ -3,6 +3,8 @@
 // Import Everything from Base.
 // This includes Bootstrap
 @import '../../../source/_patterns/00-protons/variables';
```
Comment (illepic) [inline apps/pl/scss/_scss2json.scss]: Noting here that we're going to update scss `includePaths` to make this much nicer.

File: package.json [hunk 15]
```diff
@@ -22,7 +22,7 @@
     "build:pl": "NODE_ENV='production' webpack --config ./apps/pl/webpack.pl.prod.js",
     "dev:drupal": "NODE_ENV='development' webpack --watch --progress --config ./apps/drupal/webpack.drupal.dev.js",
     "build:drupal": "NODE_ENV='production' webpack --config ./apps/drupal/webpack.drupal.prod.js",
-    "dev:grav": "NODE_ENV='development webpack --watch --progress --config ./apps/grav/webpack.grav.dev.js",
+    "dev:grav": "NODE_ENV='development' webpack --watch --progress --config ./apps/grav/webpack.grav.dev.js",
```
Comment (illepic) [inline package.json]: 👍

File: source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+@import '~font-awesome/scss/mixins';
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss]: Should these already be in the protons non-printing export?

File: source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+@import '~font-awesome/scss/variables';
+
+.svgicon-demo-0 {
+  @include sprite('chevron-right');
+  height: 40px;
+  width: 40px;
+}
+
+.svgicon-demo-1 {
+  @include sprite(
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss]: I know that `sprite()` is auto-generated, but that name is sooooo easily stepped-on...

File: source/_patterns/01-atoms/svgicon/demo/svgicons.twig [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+        .svgicon-demo-0 {
+          @include sprite('chevron-right');
+          height: 40px;
+          width: 40px;
+        }
+      </code>
+    </pre>
+      <div class="d-flex justify-content-between mb-3">
+        {% for icon in scssIcons %}
+        <div class="icon--{{ icon }}">
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/svgicons.twig]: Make these `<i>` tags. It's just a convention that's out there.

File: source/_patterns/01-atoms/svgicon/demo/svgicons.twig [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+        </div>
+        {% endfor %}
+      </div>
+
+    <h3>Changing colors with the Sass mixin</h3>
+    <p>To allow an SVG to be colorized, add <code>var:color.fill="#000"</code> to all path elements on the raw SVG. This defaults the color to black, but allows an override when using the mixin. <b class="text-danger">Warning:</b> this does involve altering the raw SVG which will need to be be updated with any revisions to the source file.</p>
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/svgicons.twig]: I want a table on this page with "techniques" per row, and "features" as columns. That way we can quickly see "how do I color an SVG that was included only via html class?" or "How do I include a full SVG element and also colorize it?"

File: source/_patterns/01-atoms/svgicon/demo/svgicons.twig [hunk 20]
```diff
@@ -0,0 +1,110 @@
+{% extends '@atoms/grid/_grid-1-up.twig' %}
+  {% set container = true %}
+  {% block column_1 %}
+
+    <h1>SVG Icons</h1>
+    <p class="lead">Store your SVG files in the SVG subdirectory to have them made available via the <code>_svgicon.twig</code> partial or using a Sass mixin. SVGs are rendered utilizing their filename as a variable.</p>
+
+    <h2>Using Twig to render SVG elements</h2>
```
Comment (illepic) [inline source/_patterns/01-atoms/svgicon/demo/svgicons.twig]: Need code example.

File: source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss (hunk 16 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss]: protons/_base.scss imports the font-awesome/scss/font-awesome. Those two lines were copied over from the existing icon scss file.

File: .gitignore [hunk 21]
```diff
@@ -3,8 +3,11 @@ dist/
 source/_data/env.json
 source/_patterns/01-atoms/icon/font/
 source/_patterns/01-atoms/icon/scss/_icon-map-generated.scss
+source/_data/scssVariables.json
```
[the remark this answers is not in the document]
Comment (dslemay) [reply .gitignore]: Removed this from gitignore in ce6105d

File: source/_patterns/01-atoms/svgicon/demo/svgicons.twig (hunk 20 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/demo/svgicons.twig]: Added in 0c6b4b6.

File: apps/pl/scss/_scss2json.scss (hunk 14 above)
Comment (dslemay) [reply apps/pl/scss/_scss2json.scss]: Is that something to add to this PR or in the future?

Comment (illepic) [reply apps/pl/scss/_scss2json.scss]: No need in this PR. I'm going to do some pretty large Sass architecture cleanup soon.

File: source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss (hunk 16 above)
Comment (illepic) [reply source/_patterns/01-atoms/svgicon/demo/_svgicon-demo.scss]: I'll hit this in Sass architecture refactor.

File: source/_patterns/01-atoms/svgicon/demo/svgicons.twig (hunk 19 above)
Comment (dslemay) [reply source/_patterns/01-atoms/svgicon/demo/svgicons.twig]: Added initial implementation in 930b025

**Comment by LawlietBlack** [comment]
I have a new commit, not sure how to add to this(?)

**Comment by LawlietBlack** [comment]
@illepic I see a note about the sprite method, can/should I change that?

**Comment by illepic** [comment]
Here's a quick recap of the updates/fixes that need to happen before this can be merged:

image

**Comment by illepic** [comment]
MERGING! Great work everyone!!

**Pull request merged**